nds/nds_power.py

```diff
@@ -87,6 +89,111 @@ def stop_profiling(self):
             print("Profiling stopped")
 
 
+def split_and_strip(str, delimiter):
+    return [s.strip() for s in str.split(delimiter) if s.strip()]
+
+
+def parse_query_content(query_content):
+    """
+    Parse query content to identify setup, benchmark, and cleanup sections.
+    
+    Args:
+        query_content (str): The full query content potentially containing timing tags.
+        
+    Returns:
+        dict: A dictionary with keys:
+            - 'query_tpl': Query name template. This will be updated based on the query type and index.
+            - 'setup': SQL string before '-- start benchmark'
+            - 'benchmark': SQL string between '-- start benchmark' and '-- end benchmark'
+            - 'cleanup': SQL string after '-- end benchmark'
+    """
+    lines = split_and_strip(query_content, '\n')
+    head = lines[0]
+    lines = lines[1:]  # Exclude the head line
+    
+    setup_lines = []
+    benchmark_lines = []
+    cleanup_lines = []
+    
+    current_section = 'init'
+    
+    for line in lines:
+        line_stripped = line.strip()
+        if not line_stripped:
+            continue
+        if line_stripped.startswith('-- end query'):
+            break
+
+        # Transitions allowed:
+        # init -> (setup) -> benchmark -> (cleanup -> done)
+        # All other transitions are invalid.
+
+        if line_stripped == '-- start setup':
+            if current_section != 'init':
+                raise RuntimeError("The setup section must be the first section if it exists.")
```

Comment:
For easier debugging of statements like this, I always suggest to include the unexpected value as welll


```suggestion
                raise RuntimeError(f"Init expected, actual section {current_section}. The setup section must be the first section if it exists.")
```

Reply:
Fixed as suggested.